### PR TITLE
Mandatory notes field in the "Send a referral" form.

### PR DESCRIPTION
**Pull request by ian-leggett**
## Description of change
The notes text area field should now be a mandatory field in the "Send a referral" form.

## Test instructions
Go to send a referral by:
1. Click a company
2. Remove the `/activity` at the end of the url
3. Add `/send-referral` at the end of the url
4. Try to submit with no notes added - you should see the error
5. Add notes and remove them - you should see the error

_What should I see?_

## Screenshots
Screenshot 2020-03-17 at 16 50 33

## Checklist

[//]: # "When submitting a PR make sure the code review guidelines have been satisfied.
https://github.com/uktrade/data-hub-frontend/blob/master/docs/Code%20review%20guidelines.md"

- [x] Has the branch been rebased to master?
- [x] Automated tests (Any of the following when applicable: Unit, Functional or Acceptance)
- [x] Manual compatibility testing (Browsers: Chrome, Firefox, IE11, Safari)

### CODE REVIEW

File: src/client/components/ValidatedTextArea.jsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    validationError,
+    id,
+    validator,
+    error,
+    ...props
+  }) => (
+    <TextArea
+      {...props}
+      meta={{ touched: true, error: validationError || error }}
+      onChange={(e) => {
```
Comment (web-bert) [inline src/client/components/ValidatedTextArea.jsx]: This validation should not happen on change but on submit as per the guidelines: https://design-system.service.gov.uk/patterns/validation/

Comment (mforner13) [reply src/client/components/ValidatedTextArea.jsx]: @web-bert It seems they're talking more about not validating the text as they enter it, e.g. for a character account. They mention not validating as the user is typing, but this validation would only happen when they delete what's in the field. But I see it's a fine line!

File: src/apps/companies/apps/referrals/send-referral/client/SendReferralForm.jsx [hunk 2]
```diff
@@ -143,14 +147,17 @@ const SendReferralForm = ({
         Subject
       </ValidatedInput>
       <br />
-      <Label>
+      <StyledValidatedTextArea
+        hint="For example the details of previous conversations with the company, or
```
Comment (mforner13) [inline src/apps/companies/apps/referrals/send-referral/client/SendReferralForm.jsx]: The new design has this content for the hint text - just to save us doing this in the future PR :) 
Screenshot 2020-03-18 at 15 38 58

File: src/apps/companies/apps/referrals/send-referral/client/SendReferralForm.jsx [hunk 3]
```diff
@@ -143,14 +147,17 @@ const SendReferralForm = ({
...
+          a description of the purpose or opportunity this referral is about"
+        id="notes"
```
Comment (mforner13) [inline src/apps/companies/apps/referrals/send-referral/client/SendReferralForm.jsx]: As this id is what the notes state will be stored under in the store, under the ValidatedTextArea portion of state, 'referralNotes' might make it a bit easier to recognise.

File: test/functional/cypress/specs/companies/referrals/send-referral-spec.js [hunk 4]
```diff
@@ -106,10 +110,29 @@ describe('Send a referral form', () => {
       () => {
         it('should display error message', () => {
           selectTypeahead('Adviser', 'shawn')
+          cy.get(selectors.companySendReferral.notesField)
+            .click()
+            .type('Example notes')
```
Comment (mforner13) [inline test/functional/cypress/specs/companies/referrals/send-referral-spec.js]: Very minor but, should it be 'Example subject' ?

File: test/functional/cypress/specs/companies/referrals/send-referral-spec.js [hunk 5]
```diff
@@ -106,10 +110,29 @@ describe('Send a referral form', () => {
...
+    )
+
+    context(
+      'when "Continue" button is clicked without specifying notes',
+      () => {
+        it('should display error message', () => {
+          selectTypeahead('Adviser', 'shawn')
+          cy.get(selectors.companySendReferral.subjectField)
+            .click()
+            .type('Example subject')
```
Comment (mforner13) [inline test/functional/cypress/specs/companies/referrals/send-referral-spec.js]: Think this is mixed up with the test above - should be 'Example notes'

File: src/client/components/ValidatedTextArea.jsx (hunk 1 above)
Comment (mforner13) [reply src/client/components/ValidatedTextArea.jsx]: This on change validation also allows the error message to go when a user types in some notes, which I think Rogier preferred to only finding out if you've satisfied the error when you submit the form.

Comment (web-bert) [reply src/client/components/ValidatedTextArea.jsx]: This is the pertinent bit for me:

> Do not validate when the user moves away from a field. Wait until they try to move to the next part of the service - usually by clicking the ‘continue’ or ‘submit’ button at the bottom of the page.

Comment (rafenden) [reply src/client/components/ValidatedTextArea.jsx]: This is a new pattern and I would advise to not go against the way current forms are working, so we can provide consistent experience.

File: src/apps/companies/apps/referrals/send-referral/client/SendReferralForm.jsx (hunk 3 above)
Comment (ian-leggett) [reply src/apps/companies/apps/referrals/send-referral/client/SendReferralForm.jsx]: Good spot, all updated 👍

File: src/client/components/ValidatedTextArea.jsx (hunk 1 above)
Comment (ian-leggett) [reply src/client/components/ValidatedTextArea.jsx]: I have now removed the new validation behaviour, the form is now consistent and behaves in the same way as other forms on DH.

File: src/apps/companies/apps/referrals/send-referral/client/SendReferralForm.jsx [hunk 6]
```diff
@@ -56,20 +57,23 @@ const SendReferralForm = ({
...
+  const emptyNotesError = emptyFields.includes('notes')
 
   return (
     <form
       onSubmit={(event) => {
         event.preventDefault()
         const emptyFields = [
-          !event.target.subject.value && 'subject',
+          (!event.target.subject.value ||
+            event.target.subject.value.length >= 255) &&
```
Comment (web-bert) [inline src/apps/companies/apps/referrals/send-referral/client/SendReferralForm.jsx]: Could this be simpler as `event.target.subject?.value?.length > 255 && 'subject'`

File: src/apps/companies/apps/referrals/send-referral/client/SendReferralForm.jsx [hunk 7]
```diff
@@ -129,28 +133,28 @@ const SendReferralForm = ({
...
-          value.length
-            ? value.length > 255 && 'Subject must be 255 characters or less'
-            : 'Enter a subject for the referral'
-        }
+      <InputField
+        meta={{
+          touched: true,
+          error:
+            emptySubjectError &&
+            'Enter a subject for the referral (Max 255 characters)',
```
Comment (web-bert) [inline src/apps/companies/apps/referrals/send-referral/client/SendReferralForm.jsx]: Might be a good idea to put the `255` in a `const` and re-use in this message and the validator so it's one thing to change if we want to alter the max length

**Pull request merged**